### PR TITLE
CY-1834 Update provision.sh to use docker

### PR DESCRIPTION
**Pull request by mabuaisha**

### CODE REVIEW

File: packaging/plugins/provision.sh [hunk 1]
```diff
@@ -14,50 +16,39 @@ function print_plugins_params() {
     done
 }
 
-function install_dependencies(){
+function install_windows_dependencies(){
+    # Prepare virtualenv in windows machine
     echo "## Installing necessary dependencies"
+    pip install virtualenv
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: please pin to a version, so it will be stable

File: packaging/plugins/provision.sh [hunk 2]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+}
+
+function override_constraints_file() {
+    if [ ! -z "$CONSTRAINTS_FILE" ];then
+        cp -rf $CONSTRAINTS_FILE $PLUGIN_PATH/
+    fi
+}
+
+function clean_docker_resources() {
+  # This will clean all exited docker containers
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: If it's stuck?
maybe use a specific name

File: packaging/plugins/provision.sh [hunk 3]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+
+function get_wagon_builder_name() {
+	echo "## get wagon builder name"
+	WAGON_BUILDLER=$(basename $WAGON_BUILDLER_REPO | cut -f1 -d".")
 }
 
+function wagon_create_package(){
+
+    echo "## wagon create package"
+    # Checkout to the plugin
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: This comment is redundant, it's clear enough

File: packaging/plugins/provision.sh [hunk 4]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+    checkout_plugin
+
+    # Check if constraint file is provided or not
+    override_constraints_file
+
+    # This will generate a wagon file and dump it to the current plugin name
+    # directory, this should work for all linux image but for Redhat we need
+    # to build it locally since it needs subscription account
+    echo "## echo build wagon package using docker"
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: Not sure about the next two echo-s

File: packaging/plugins/provision.sh [hunk 5]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+function clean_docker_resources() {
+  # This will clean all exited docker containers
+  docker rm $(docker ps -a -f status=exited -q)
+
+  # Remove Docker image
+  docker rmi $IMAGE_NAME --force
+}
+
+function get_wagon_builder_name() {
+	echo "## get wagon builder name"
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: This comment is redundant

File: packaging/plugins/provision.sh [hunk 6]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+    override_constraints_file
+
+    # This will generate a wagon file and dump it to the current plugin name
+    # directory, this should work for all linux image but for Redhat we need
+    # to build it locally since it needs subscription account
+    echo "## echo build wagon package using docker"
+    echo "clone $WAGON_BUILDLER_REPO"
+    git clone $WAGON_BUILDLER_REPO
+    # Get the name of the wagon builder project to navigate inside it
+    get_wagon_builder_name
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: why not return to here?
also you can just move the impl to here, it's only used here

File: packaging/plugins/provision.sh [hunk 7]
```diff
@@ -67,17 +58,83 @@ function wagon_create_package(){
...
+             IMAGE_NAME="cloudify-$PLUGIN_PLATFORM-wagon-builder"
+             if [[ $PLUGIN_PLATFORM == "redhat"* ]]; then
+                  docker build -t $IMAGE_NAME --build-arg USERNAME=$REL_SUB_USERNAME --build-arg PASSWORD=$REL_SUB_PASSWORD .
+             else
+                  IMAGE_NAME=$DOCKER_ACCOUNT/$IMAGE_NAME
+             fi
+         popd
+    popd
 
+    docker run -v $PLUGIN_PATH:/packaging $IMAGE_NAME
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: Add a closing echo to mark the end of processes across this script

File: packaging/plugins/provision.sh [hunk 8]
```diff
@@ -88,14 +145,34 @@ PLUGIN_S3_FOLDER=$7
...
+    echo "# Docker Builder is running....."
+    PLUGIN_PLATFORM=${10}
+    WAGON_BUILDLER_REPO=${11}
+    WAGON_BUILDLER_BRANCH=${12}
+    DOCKER_ACCOUNT=${13}
+    REL_SUB_USERNAME=${14}
+    REL_SUB_PASSWORD=${15}
+    DOCKER_BUILDER=true
+else
+  echo "# Vagrant Builder is running....."
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: I think you can remove it completely

Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Done

File: packaging/plugins/provision.sh (hunk 7 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Done

File: packaging/plugins/provision.sh (hunk 6 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Done

File: packaging/plugins/provision.sh (hunk 5 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: done

File: packaging/plugins/provision.sh (hunk 4 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Jus kept this one and remove others

File: packaging/plugins/provision.sh (hunk 3 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: removed

File: packaging/plugins/provision.sh (hunk 1 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Done

File: packaging/plugins/provision.sh (hunk 2 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Sure will update that

**Comment by mabuaisha** [comment]
@tyacbovi I updated the code based on your comments and fixed all the stuff that you pointed to

File: packaging/plugins/provision.sh [hunk 9]
```diff
@@ -124,7 +119,8 @@ function wagon_create_package(){
          popd
     popd
 
-    docker run -v $PLUGIN_PATH:/packaging $IMAGE_NAME
+    docker run --name $CONTAINER_NAME -v $PLUGIN_PATH:/packaging $IMAGE_NAME
+    echo "## echo build wagon package created successfully"
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: Well you don't need to write echo in the echo message :)

File: packaging/plugins/provision.sh [hunk 10]
```diff
@@ -156,8 +152,6 @@ if [ "$#" -gt 9 ]; then
     REL_SUB_USERNAME=${14}
     REL_SUB_PASSWORD=${15}
     DOCKER_BUILDER=true
```
Comment (tyacbovi) [inline packaging/plugins/provision.sh]: So DOCKER_BUILDER is not needed anymore

File: packaging/plugins/provision.sh (hunk 9 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Removed

File: packaging/plugins/provision.sh (hunk 10 above)
Comment (mabuaisha) [reply packaging/plugins/provision.sh]: Well, you are right, but building wagon for windows in near future is need to be supported and already handled here but never used by our build system, some client use windows and I you noticed that script you will see that vagrant config for windows still there

**Pull request merged**